stl/inc/xutility

```diff
@@ -2215,19 +2213,18 @@ namespace ranges {
 
     // CONCEPT ranges::range
     template <class _Rng>
-    concept _Range_impl = requires(_Rng&& __r) {
-        _RANGES begin(static_cast<_Rng&&>(__r));
-        _RANGES end(static_cast<_Rng&&>(__r));
+    concept range = requires(_Rng& __r) {
+        _RANGES begin(__r);
+        _RANGES end(__r);
     };
 
-    template <class _Rng>
-    concept range = _Range_impl<_Rng&>;
+    // VARIABLE TEMPLATE ranges::enable_safe_range
+    template <range>
+    inline constexpr bool enable_safe_range = false;
 
-    // CONCEPT ranges::_Forwarding_range
-    // clang-format off
+    // CONCEPT ranges::safe_range
     template <class _Rng>
-    concept _Forwarding_range = range<_Rng> && _Range_impl<_Rng>;
-    // clang-format on
+    concept safe_range = range<_Rng> && (is_lvalue_reference_v<_Rng> || enable_safe_range<remove_cvref_t<_Rng>>);

```

Comment:
@CaseyCarter I have some reservations about this line. Technically `is_lvalue_reference_v` is not the opposite of `is_rvalue_reference_v`, so I would have actually expected `(!is_rvalue_reference_v<_Rng> || enable_safe_range<remove_cvref_t<_Rng>>)`

Reply:
But if for example `_Rng` was `std::vector`, `safe_range<_Rng>` would be `true` even though it's iterators are tied to lifetime of it (when `std::vector` gets destroyed it's iterators get invalidated). lvalue references don't own the object, so after reference is destroyed the range it referenced may still exist and have valid iterators.

Reply:
I think I didnt make my point well. The concept `safe_range` differs slightly from the requirements on `ranges::begin` and friends. `safe_range` requires `is_lvalue_reference_v || enable_safe_range` whereas the others all require `!is_rvalue_reference_v || enable_safe_range`.

This subtle difference smells like a bug and I would expect those requirements to match each other. Your point is definitely correct, that `!is_rvalue_reference_v ` would allow copies that end up as dangling references so maybe the requirement on `ranges::begin` should be adopted.

Nevertheless, this feels inconsitent and should be verified by the ranges expert

Reply:
I went ahead and defined an internal concept `_Safe_range_like` that encapsulates the requirement and is used o constraint the template